internal/oidc/oauth2.go

```diff
@@ -0,0 +1,152 @@
+	var store = NewStore(configuration)
+
+	var oidcConfig = new(compose.Config)
+
+	privateKey, err := utils.ParseRsaPrivateKeyFromPemStr(configuration.IssuerPrivateKey)
+	if err != nil {
+		panic(err)
+	}
+
+	oauth2 := compose.ComposeAllEnabled(oidcConfig, store, []byte(configuration.HMACSecret), privateKey)
+
+	// TODO: Add paths for UserInfo, Flush, Logout.
+
+	// TODO: Add OPTIONS handler.
+	router.GET(wellKnownPath, autheliaMiddleware(WellKnownConfigurationHandler))
+
+	router.GET(consentPath, autheliaMiddleware(ConsentGet))
+
+	router.POST(consentPath, autheliaMiddleware(ConsentPost))
+
+	router.GET(jwksPath, autheliaMiddleware(JWKsGet(&privateKey.PublicKey)))
+
+	router.GET(authorizePath, autheliaMiddleware(middlewares.NewHTTPToAutheliaHandlerAdaptor(AuthorizeEndpointGet(oauth2))))
+
+	// TODO: Add OPTIONS handler.
+	router.POST(tokenPath, autheliaMiddleware(middlewares.NewHTTPToAutheliaHandlerAdaptor(tokenEndpoint(oauth2))))
+
+	router.POST(introspectPath, autheliaMiddleware(middlewares.NewHTTPToAutheliaHandlerAdaptor(introspectEndpoint(oauth2))))
+
+	// TODO: Add OPTIONS handler.
+	router.POST(revokePath, autheliaMiddleware(middlewares.NewHTTPToAutheliaHandlerAdaptor(revokeEndpoint(oauth2))))
+}
+
+// A session is passed from the `/auth` to the `/token` endpoint. You probably want to store data like: "Who made the request",
+// "What organization does that person belong to" and so on.
+// For our use case, the session will meet the requirements imposed by JWT access tokens, HMAC access tokens and OpenID Connect
+// ID Tokens plus a custom field.
+
+// newSession is a helper function for creating a new session. This may look like a lot of code but since we are
+// setting up multiple strategies it is a bit longer.
+// Usually, you could do:
+//
+//  session = new(fosite.DefaultSession)
+func newSession(ctx *middlewares.AutheliaCtx, scopes fosite.Arguments, audience fosite.Arguments) *openid.DefaultSession {
+	session := ctx.GetSession()
+
+	extra := map[string]interface{}{}
+
+	if len(session.Emails) != 0 && scopes.Has("email") {
+		extra["email"] = session.Emails[0]
+	}
+
+	if scopes.Has("groups") {
+		extra["groups"] = session.Groups
+	}
+
+	if scopes.Has("profile") {
+		extra["profile"] = map[string]string{
+			"name": session.DisplayName,
+		}
+	}
+
+	/*
+		TODO: Adjust auth backends to return more profile information.
+		It's probably ideal to adjust the auth providers at this time to not store 'extra' information in the session
+		storage, and instead create a memory only storage for them.
+		This is a simple design, have a map with a key of username, and a struct with the relevant information.
+	*/
+
+	oidcSession := newDefaultSession(ctx)
+	oidcSession.Claims.Extra = extra
+	oidcSession.Claims.Subject = session.Username
+	oidcSession.Claims.Audience = audience
+
+	return oidcSession
+}
+
+func newDefaultSession(ctx *middlewares.AutheliaCtx) *openid.DefaultSession {
```

Comment:
Is this even used. The content of Audience makes me wondering. We should remove it if not.

Reply:
It is, it's the basis for handlers when we don't have the authorize request to obtain the granted scopes/audience, specifically the token endpoint. We also use it as a basis for the newSession() for consistency sake. I do think we need to invest some time in this area though.

I'm thinking this is likely something we should do after we figure out storage, as I believe the memory storage example allows you to do `storage.IDTokens["token"].GetSession()`. I can test this in making some alterations per your other comments.